### DISCUSSION TITLE
Exif: Questions regarding SetValue and charset

### DISCUSSION THREAD

**Discussion by BareGeode** [dlemstra/Magick.NET#1715] [Help] [answered]
So, I am trying to save an Exif UserComment to a .webp image like this:

```
using var image = new MagickImage("apple.webp");
var profile = image.GetExifProfile();
profile.SetValue(ExifTag.UserComment, Encoding.UTF8.GetBytes("Wonderful wonderful little cømment"));
image.SetProfile(profile);
image.Write("apple.webp");
```

It does work (partially), but I have some issues. First of all, I think that I'm doing SetValue wrong somehow, since I'm not setting TValueType for SetValue. When I try to do that, like this:
`profile.SetValue<string>(ExifTag.UserComment, "Wonderful wonderful little cømment");`
I get that "Argument 1: cannot convert from 'ImageMagick.ExifTag<byte[]>' to 'ImageMagick.ExifTag\<string>". Not sure how to do this correctly.

Secondly, with the way I do it, if I then open the image in GIMP, I can see that the charset isn't set correctly since it says "charset=InvalidCharsetId l wonderful little cømment" for the comment of the image. Probably related to how I do SetValue, or what am I doing wrong?

**Comment by dlemstra** [comment, marked as answer]
The format is a byte array because the value is an "undefined type" so you will need to write the bytes yourself. I did find this PR that contained some information about that you should do: https://github.com/SixLabors/ImageSharp/issues/1919.

**Comment by BareGeode** [comment]
Thanks for the reply. I went down a rabbit hole and have come to the conclusion that everyone just adds 8 null bytes before the text (i.e. set it to the "undefined character code" per the standard), even though there is a Unicode character code that specifies UTF-8 per version 3.0 of the EXIF standard. If I write the Unicode character code as one would think one should, then most tools either read it incorrectly ("坯湤敲晵氠睯湤敲晵氠汩瑴汥⁣쎸浭敮" in ExifTool) or read the character code as if it were a part of the text ("charset=Unicode Wonderful wonderful little cømment" in GIMP). Magick.NET reads the character code regardless, but it's not visible when converted to a string if set to the Unicode charset, since that's just null bytes which get ignored. If set to Unicode, the UserComment is read like this: "UNICODEWonderful wonderful little cømment". So, I'll just add 8 null bytes before the comment x)
